### PR TITLE
Optimized farm.stop and worktimeout in stratum

### PR DESCRIPTION
**Pull request by smurfy**
* More fixes for farm.stop
* Re-enabled api event for miner restart
* On disconnect shutdown all mining to save wasted power
* Start worktimeout in stratum right after a successful connection. So if connection is successfull but something else failes and you never get work, this triggers a reconnect after the timeout is reached 

Disclaimer:
farm.stop()  still can cause driver resets, i modified my simulator to disconnect after a while but never had a driver reset with this test after 2h of simulating. But i had driver resets by stopping the miner, which should do the same so i do not know what causes the driver reset.

### CODE REVIEW

**Comment by jean-m-cyr** [comment]
@smurphy Yes. much cleaner.

**Comment by smurfy** [comment]
Fixes: #672 #740

**Comment by smurfy** [comment]
Hm, ok, need to test on linux.
On windows it works.

````
  m  22:28:37|main    |  Speed  26.13 Mh/s    gpu/0 26.13 58C 47%   [A0+0:R0+0:F0] Time: 00:01
  m  22:28:42|main    |  Speed  26.30 Mh/s    gpu/0 26.30 58C 47%   [A0+0:R0+0:F0] Time: 00:02
  m  22:28:47|main    |  Speed  26.14 Mh/s    gpu/0 26.14 58C 47%   [A0+0:R0+0:F0] Time: 00:02
  X  22:28:51|stratum |  Read response failed: The semaphore timeout period has expired
  i  22:28:51|stratum |  Disconnected from eu2.ethermine.org
  i  22:28:51|stratum |  Shutting down miners...
  i  22:28:51|stratum |  Retrying in 3 ...
  m  22:28:52|main    |  not-connected
  i  22:28:52|stratum |  Retrying in 2 ...
  i  22:28:53|stratum |  Retrying in 1 ...
  X  22:28:54|stratum |  Could not resolve host eu2.ethermine.org:4444,  No such host is known
  i  22:28:54|stratum |  Disconnected from eu2.ethermine.org
  i  22:28:54|stratum |  Retrying in 3 ...
  i  22:28:55|stratum |  Retrying in 2 ...
````

**Comment by jean-m-cyr** [comment]
Simple rig. 4 nvidia 1060. Linux 16.4. SSL enabled. It pretty much happens every time hee. Let me know if there's debug data I can collect.

**Comment by jean-m-cyr** [comment]
Been seeing this for a while. Maybe it's a local problem, but if I go back to pre-pool manager it goes away.

**Comment by smurfy** [comment]
maybe i do not understand what you mean by "Disconnect wire", or what the problem is you are having.
It is my understanding, that even after unplugging the ethernet cable it still mines and only disconnects after the no-work timeout triggers.

I tested on windows and by unplugging the cable (Takes longer to detect disco, around 10-20 sec) and disabling the ethernet device. (detects almost instantly).

I did some research and found nothing helpful in our case.

As i understand the documentation, ``async_read_until`` should call the handler after either data arrives or an error occurs, which it does on windows.

So i really dont understand whats happening :)

**Comment by jean-m-cyr** [comment]
After I physically pull the Ethernet cable it continues mining showing the normal 5 second interval hash rate logs, may even find and submit a few shares (without response), then after almost exactly 2 minutes the log output freezes and it won't respond to ctl-c. This is without your keep-alive patch.

**Comment by smurfy** [comment]
ok, i understood correctly then :)

Regarding crash/freeze try commenting this line out ``m_socket.close``  i think pre stratum version had a comment regarding crashes on linux.

https://github.com/smurfy/ethminer/blob/b7558672f944b988a5e95e63762315eab695bf62/libpoolprotocols/stratum/EthStratumClient.cpp#L180

> This is without your keep-alive patch.

while the above fixes maybe the crash, maybe the keep-alive addition fixes your not detecting the disconnect.

**Comment by smurfy** [comment]
The old code almost exclusively called ``reconnect``, which had the ``m_socket.close`` call commented out. 
Since we now call ``disconnect`` (which has and had the call) and manager reconnects,
commenting out the call on linux could be the fix.

I'm also not sure if we comment out the close call, if the socket actually disconnects or not. 

For reference: 
https://github.com/ethereum-mining/ethminer/blob/v0.13.0/libstratum/EthStratumClient.cpp#L106

**Comment by smurfy** [comment]
Alternatives of commenting out. (if it fixes the problem)

Call it with error code and check if it still crashes.
````
boost::system::error_code ec;
m_socket.close(ec);
if (ec)
{
  // An error occurred.
}
````

OR

Call ``m_socket.shutdown()``  before the close.

**Comment by jean-m-cyr** [comment]
commenting out the close I get this after 3 minutes:

  m  19:22:20|ethminer|  Speed  83.36 Mh/s    gpu/0 21.61  gpu/1 18.16  gpu/2 21.80  gpu/3 21.80  [A1+0:R0+0:F0] Time: 00:03
  m  19:22:25|ethminer|  Speed  83.36 Mh/s    gpu/0 21.57  gpu/1 18.16  gpu/2 21.83  gpu/3 21.80  [A1+0:R0+0:F0] Time: 00:03
  m  19:22:30|ethminer|  Speed  83.36 Mh/s    gpu/0 21.57  gpu/1 18.16  gpu/2 21.83  gpu/3 21.80  [A1+0:R0+0:F0] Time: 00:03
  m  19:22:35|ethminer|  Speed  83.40 Mh/s    gpu/0 21.61  gpu/1 18.16  gpu/2 21.83  gpu/3 21.80  [A1+0:R0+0:F0] Time: 00:03
  m  19:22:40|ethminer|  Speed  83.32 Mh/s    gpu/0 21.57  gpu/1 18.13  gpu/2 21.83  gpu/3 21.80  [A1+0:R0+0:F0] Time: 00:03
  m  19:22:45|ethminer|  Speed  83.32 Mh/s    gpu/0 21.57  gpu/1 18.13  gpu/2 21.83  gpu/3 21.80  [A1+0:R0+0:F0] Time: 00:03
  m  19:22:50|ethminer|  Speed  83.32 Mh/s    gpu/0 21.57  gpu/1 18.13  gpu/2 21.83  gpu/3 21.80  [A1+0:R0+0:F0] Time: 00:04
  ℹ  19:22:51|stratum |  No new work received in 180  seconds.
^[[101m  ✘  19:22:52|main    |  Exception thrown in Worker thread: write: protocol is shutdown
  m  19:22:55|ethminer|  Speed  83.25 Mh/s    gpu/0 21.53  gpu/1 18.13  gpu/2 21.80  gpu/3 21.80  [A1+0:R0+0:F0] Time: 00:04
  m  19:23:00|ethminer|  Speed  83.32 Mh/s    gpu/0 21.57  gpu/1 18.13  gpu/2 21.80  gpu/3 21.83  [A1+0:R0+0:F0] Time: 00:04

**Comment by jean-m-cyr** [comment]
Above was with SSL. Did same, with close() commented out, on standard TCP connection and work timeout and switchover works fine.

**Comment by jean-m-cyr** [comment]
Nope, still broke:

Same error:
  m  16:08:51|ethminer|  Speed  69.34 Mh/s    gpu/0 19.46  gpu/1 15.66  gpu/2 17.37  gpu/3 16.84  [A0+0:R0+0:F0] Time: 00:00
  m  16:08:56|ethminer|  Speed  83.26 Mh/s    gpu/0 21.44  gpu/1 18.14  gpu/2 21.81  gpu/3 21.86  [A0+0:R0+0:F0] Time: 00:00
  â  16:08:56|stratum |  No new work received in 30  seconds.
  â  16:08:59|main    |  Exception thrown in Worker thread: write: protocol is shutdown
  m  16:09:01|ethminer|  Speed  83.36 Mh/s    gpu/0 21.55  gpu/1 18.14  gpu/2 21.81  gpu/3 21.86  [A0+0:R0+0:F0] Time: 00:00

Also got the same on response timeout

  m  16:07:43|ethminer|  Speed  71.51 Mh/s    gpu/0 19.71  gpu/1 15.99  gpu/2 18.09  gpu/3 17.72  [A0+0:R0+0:F0] Time: 00:00
  m  16:07:48|ethminer|  Speed  83.20 Mh/s    gpu/0 21.44  gpu/1 18.14  gpu/2 21.81  gpu/3 21.81  [A0+0:R0+0:F0] Time: 00:00
  â¹  16:07:48|cuda-3  |  Nonce 0x2a68d284c24bcd07 submitted to us1-etc.ethermine.org
  â  16:07:50|stratum |  No no response received in 2 seconds.
  m  16:07:53|ethminer|  Speed  83.31 Mh/s    gpu/0 21.55  gpu/1 18.14  gpu/2 21.81  gpu/3 21.81  [A0+0:R0+0:F0] Time: 00:00
  â  16:07:57|main    |  Exception thrown in Worker thread: write: protocol is shutdown
  â¹  16:07:57|cuda-1  |  Nonce 0x2a68d084c9544ecb submitted to us1-etc.ethermine.org

Seems that calls to ->disconnect() mess things up. I don't think the the fault happens where we think, but instead in some asynchronous handler... maybe. I tried stopping the io_service before the close instead of after but that didn't help. Is there a thread that continues to send even when disconnected?

**Comment by smurfy** [comment]
> Is there a thread that continues to send even when disconnected?

well we need to add more try and catch

File: libpoolprotocols/stratum/EthStratumClient.cpp [hunk 1]
```diff
@@ -173,13 +173,21 @@ void EthStratumClient::disconnect()
...
+	try {
+		if (m_secureMode != StratumSecure::NONE) {
+			boost::system::error_code sec;
+			m_securesocket->shutdown(sec);
+		}
 
-	m_io_service.stop();
-	m_socket->close();
+		m_io_service.stop();
+#ifdef _WIN32
```
Comment (jean-m-cyr) [inline libpoolprotocols/stratum/EthStratumClient.cpp]: This is not necessary. The socket close is ok in Linux if you move the m_io_service.stop() after it. Makes sense!

Comment (jean-m-cyr) [reply libpoolprotocols/stratum/EthStratumClient.cpp]: Doesn't help for SSL though.

**Comment by smurfy** [comment]
I will try to get this fixed using a virtual machine. I cant plug my cable on my rig all the time :)

**Comment by jean-m-cyr** [comment]
How do you run video card code in a VM?

**Comment by smurfy** [comment]
i don't, i comment out the opencl checks and keep just the stratum stuff around.
Hopefully this works. i also set the work timeout to 10 seconds or so. That way i was able to test some stuff before.

**Comment by AndreaLanfranchi** [comment]
> I will try to get this fixed using a virtual machine. I cant plug my cable on my rig all the time :)

I think it may result easier to ...

On linux 

```
ifdown eth0
ifup eth0
```

On windows enable or disable the network interface.

No need to unplug/replug the cable each time.

**Comment by jean-m-cyr** [comment]
One thing I was looking at. I think we continuously have a pending read against the pool ssl socket. When you call shutdown(), the readhandler should get called with an error code. Are we handling that correctly, or is it somehow generating a further write?

**Comment by smurfy** [comment]
@jean-m-cyr well i need to do some tests, i had to recreate the sockets, because somehow i was not able to reuse them after stop/close/shutdown . Not sure why. So its possible, that the old socket (which i delete) or io_service is still there somehow.

@AndreaLanfranchi 

this does not work, at least not on windows. a disable is INSTANTLY detected, while a cable pull takes a bit to get detected. i even briefly can plug the cable and put it in a second later and the connection is still working (getting work from pool)

Not sure about linux, but i think its similar.  I think the difference is, a eth disable closes the connection CLIENT side, while a cable plug simulates/acts similarly to a SERVER close/timeout.

**Comment by jean-m-cyr** [comment]
Linux does not recognize a cable pull instantly, but the Linux desktop recognizes it in under a second. Somehow that is not percolating up to our socket!

**Comment by jean-m-cyr** [comment]
@smurfy A clue? It is the socket writes in senHashReport that are causing the "Worker thread: write: protocol is shutdown" error. Even with the if ... return at the beginning!! Also, if you change the writes to async_writes, it doesn't crash anymore, it doesn't disconnect but no crash.

**Comment by jean-m-cyr** [comment]
Another clue. If you start ethminer with the ethernet cable unplugged, it will happily retry both primary and fallback pools, till you plug in the ethernet cable at which time the console freezes and won't respond to ctl-c.

**Comment by jean-m-cyr** [comment]
More data. put a cout statement before and after the tls shutdown call and you'll see that when the cable is pulled and either the response or work timer expires, the disconnect method hangs indefinitely in shutdown!!! Any call to disconnect, regardless of caller will hang there.

**Comment by jean-m-cyr** [comment]
So, to recap, when using TLS and the ethernet cable is pulled:
- The secure layer shutdown() call will hang, and some time later the synchronous write in the report hashrate method will cause a 'write: protocol is shutdown exception'. Since disconnect doesn't set the m_disconnected bool till it's done (and it's hung in shutdown), the HR writes continue to happen.
- Changing the synchronous writes in the HR reporting method, circumvents the write exception, but I don't think that's the answer.

**Comment by aleqx** [comment]
Hi guys, I'm wondering if the issues discussed are related to a strange behaviour I've been experiencing after compiling master about 2 weeks ago. When the pool breaks connection, ethminer reconnects, subscribes and gets authorized, but then reports no new jobs for many minutes on end until it eventually times out (10-30 minutes) and reconnects again ... upon reconnection it usually works fine again, but it may happen that it goes into the same state again.

Out of 200+ miners, about a fourth to a third do so, wasting about 30 minutes of mining time until becoming operational.

Of the above sounds familiar, has it been fixed in master?

p.s. I did try the current master but it gets me rejected shares continuously (something's awfully wrong).

**Comment by aleqx** [comment]
Thanks. I'll give it another go when it actually gets some accepted solutions :) ... not sure what happened lately, starting with 0.13dev there has been quite a few versions with issues -- i'm not sure which one I can trust (i keep going back to my modded 0.12, which implements the asio connection fix and removes all software evaluations).

**Pull request merged**